### PR TITLE
Use LoopbackServer for HttpWebRequestTests

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes https://github.com/dotnet/corefx/issues/22849
Fixes https://github.com/dotnet/corefx/issues/37430

Enabling a few tests which weren't running.

I left the synchronous tests as they were as I wasn't sure what the pattern for testing with the  LoopBackServer synchronously is.

### CODE REVIEW

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 1]
```diff
@@ -223,18 +223,23 @@ public void AllowReadStreamBuffering_SetTrueThenGet_ExpectTrue(Uri remoteServer)
...
+                await server.HandleRequestAsync();
+                using (WebResponse response = await getResponse)
+                using (Stream myStream = response.GetResponseStream())
+                using (var sr = new StreamReader(myStream))
+                {
+                    string strContent = sr.ReadToEnd();
+                    long length = response.ContentLength;
+                    Assert.Equal(strContent.Length, length);
+                }
+            });
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Nit: we generally use CreateClientAndServerAsync, e.g.
```C#
await LoopbackServer.CreateServerAsync(async uri =>
{
    HttpWebRequest request = WebRequest.CreateHttp(uri);
    using (WebResponse response = await request.GetResponseAsync())
    ... // etc
},
server => server.HandleRequestAsync());
```

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 2]
```diff
@@ -277,7 +282,7 @@ public async Task Headers_SetAfterRequestSubmitted_ThrowsInvalidOperationExcepti
             {
                 HttpWebRequest request = WebRequest.CreateHttp(uri);
                 Task<WebResponse> getResponse = request.GetResponseAsync();
-                await server.AcceptConnectionSendResponseAndCloseAsync();
+                await server.HandleRequestAsync();
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Why this change?

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 3]
```diff
@@ -223,18 +223,23 @@ public void AllowReadStreamBuffering_SetTrueThenGet_ExpectTrue(Uri remoteServer)
             Assert.True(request.AllowReadStreamBuffering);
         }
 
-        [Theory, MemberData(nameof(EchoServers))]
-        public async Task ContentLength_Get_ExpectSameAsGetResponseStream(Uri remoteServer)
+        [Fact]
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: This is causing us to lose testing of both http and https; now it's just validating http.  It'd be good to have both.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 4]
```diff
@@ -1098,18 +1103,24 @@ public void BeginGetRequestStream_CreatePostRequestThenCallTwice_ThrowsInvalidOp
...
-                request.BeginGetRequestStream(null, null);
+                HttpWebRequest request = HttpWebRequest.CreateHttp(url);
+
+                IAsyncResult asyncResult = request.BeginGetResponse(null, null);
+                Assert.Throws<ProtocolViolationException>(() =>
+                {
+                    request.BeginGetRequestStream(null, null);
+                });
+
+                return Task.FromResult<object>(null);
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: `return Task.CompletedTask;` ?

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 5]
```diff
@@ -1158,34 +1171,42 @@ public async Task GetRequestStreamAsync_SetPOSTThenGet_ExpectNotNull()
...
 
+        [Fact]
         public async Task GetResponseAsync_GetResponseStream_ExpectNotNull()
         {
             await LoopbackServer.CreateServerAsync(async (server, url) =>
             {
                 HttpWebRequest request = WebRequest.CreateHttp(url);
-                using (WebResponse response = await request.GetResponseAsync())
+                Task<WebResponse> getResponse = request.GetResponseAsync();
+                await server.HandleRequestAsync();
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Same CreateClientAndServerAsync comment.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 6]
```diff
@@ -1158,34 +1171,42 @@ public async Task GetRequestStreamAsync_SetPOSTThenGet_ExpectNotNull()
...
                 {
-                    Assert.NotNull(response.GetResponseStream());
+                    Assert.NotNull(myStream);
                 }
             });
         }
 
-        [Theory, MemberData(nameof(EchoServers))]
-        public async Task GetResponseAsync_GetResponseStream_ContainsHost(Uri remoteServer)
+        [Fact]
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Same http/https comment.

I'll stop commenting on these.  They apply to most of these tests.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 3 above)
Comment (ViktorHofer) [reply src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Is there a recommended pattern with the loopbackserver to test both?

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 1 above)
Comment (ViktorHofer) [reply src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Your example shows `CreateServerAsync` but you recommended to use `CreateClientAndServerAsync`. Am I missing something here?

Comment (stephentoub) [reply src/System.Net.Requests/tests/HttpWebRequestTest.cs]: That was a typo... the signature I used was for CreateClientAndServerAsync, I just copy-pasted the wrong name from your code.

**Comment by davidsh** [comment]
@ViktorHofer Are you still working on this PR?

**Comment by ViktorHofer** [comment]
Yes I'm willing to finish this but need a recommendation here: https://github.com/dotnet/corefx/pull/37451#discussion_r281274505

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 3 above)
Comment (stephentoub) [reply src/System.Net.Requests/tests/HttpWebRequestTest.cs]: > Is there a recommended pattern with the loopbackserver to test both?

You can, for example, pass in a bool with an InlineData(false) and InlineData(true), and then use that to initialize the LoopbackServer.Options passed in, e.g. `new LoopbackServer.Options { UseSsl = secure }`.

If that doesn't work for some reason, we could stick with just http.

Comment (davidsh) [reply src/System.Net.Requests/tests/HttpWebRequestTest.cs]: See example:
https://github.com/dotnet/corefx/blob/1064d0352c007fd2d8fb32d7f6e286fbb6d41264/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs#L504-L514

Comment (ViktorHofer) [reply src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Thanks!

**Comment by ViktorHofer** [comment]
@stephnetoub @davidsh for SSL I had to set `request.ServerCertificateValidationCallback = (a, b, c, d) => true;` so that the validation check for the local certificate is skipped. Is that right or am I expected to trust the certificate on my box?

**Comment by davidsh** [comment]
>@stephentoub @davidsh for SSL I had to set request.ServerCertificateValidationCallback = (a, b, c, d) => true; so that the validation check for the local certificate is skipped. Is that right or am I expected to trust the certificate on my box?

We don't want tests to rely on machine configuration (like setting a test certificate to be trusted on a dev box).  So, yes, use the callback you showed.  There are some exceptions to this rule for some detailed TLS testing we do, but for these tests we just want to exercise TLS. It's not important that we override trust for the certificate.

**Comment by ViktorHofer** [comment]
OK great, then this PR should be ready to go in. Please take another look. Thanks.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 7]
```diff
@@ -223,18 +223,26 @@ public void AllowReadStreamBuffering_SetTrueThenGet_ExpectTrue(Uri remoteServer)
...
+            var options = new LoopbackServer.Options { UseSsl = useSsl };
+
+            await LoopbackServer.CreateClientAndServerAsync(async uri =>
             {
-                string strContent = sr.ReadToEnd();
-                long length = response.ContentLength;
-                Assert.Equal(strContent.Length, length);
-            }
+                HttpWebRequest request = WebRequest.CreateHttp(uri);
+                request.ServerCertificateValidationCallback = (a, b, c, d) => true;
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: This is ok, but for future reference you can avoid the "a, b, c, d" by using the anonymous method instead of lambda syntax, e.g. `delegate { return true; }`.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 8]
```diff
@@ -1098,18 +1106,24 @@ public void BeginGetRequestStream_CreatePostRequestThenCallTwice_ThrowsInvalidOp
...
             {
-                request.BeginGetRequestStream(null, null);
+                HttpWebRequest request = HttpWebRequest.CreateHttp(url);
+
+                IAsyncResult asyncResult = request.BeginGetResponse(null, null);
+                Assert.Throws<ProtocolViolationException>(() =>
+                {
+                    request.BeginGetRequestStream(null, null);
+                });
+
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Should we call request.EndGetResponse on the asyncResult and then dispose of it (after the Assert.Throws)?

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 7 above)
Comment (ViktorHofer) [reply src/System.Net.Requests/tests/HttpWebRequestTest.cs]: I mimicked what was already there: https://github.com/dotnet/corefx/blob/master/src/System.Net.Requests/tests/HttpWebRequestTest.cs#L886. Thanks for the suggestion, good to know.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 9]
```diff
@@ -1098,18 +1106,24 @@ public void BeginGetRequestStream_CreatePostRequestThenCallTwice_ThrowsInvalidOp
...
+                return Task.CompletedTask;
             });
         }
 
+        [Fact]
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: I wonder how many pther tests we have across the repo that are accidentally missing [Fact] or [Theory].  Can we run an analyzer to find other such cases?

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 10]
```diff
@@ -1209,34 +1235,48 @@ public void Range_Add_Success(Uri remoteServer)
...
-                Assert.True(strContent.Contains(RequestBody));
-            }
+                using (WebResponse response = await request.GetResponseAsync())
+                using (Stream myStream = response.GetResponseStream())
+                using (var sr = new StreamReader(myStream))
+                {
+                    string strContent = sr.ReadToEnd();
+                    Assert.True(strContent.Contains(RequestBody));
+                }
+            }, server => server.HandleRequestAsync(content: RequestBody), options);
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: This isn't right.  The previous version was using an echo server, and by the client code checking that the response contained the request body, it was thus checking that the request body was correctly sent to the server (which then echo'd it).  This change is making it so that the server simply ignores the request body and instead simply sends back the same content, without verifying that's actually what it received.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 11]
```diff
@@ -1209,34 +1235,48 @@ public void Range_Add_Success(Uri remoteServer)
...
-            response.Dispose();
+            var options = new LoopbackServer.Options { UseSsl = useSsl };
+
+            await LoopbackServer.CreateClientAndServerAsync(async uri =>
+            {
+                HttpWebRequest request = WebRequest.CreateHttp(uri);
+                request.ServerCertificateValidationCallback = (a, b, c, d) => true;
+                request.UseDefaultCredentials = true;
+
+                using WebResponse response = await request.GetResponseAsync();
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Could also be:
```C#
(await request.GetResponseAsync()).Dispose();
```
in which case we wouldn't need an unused variable.

Not a big deal, though.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 12]
```diff
@@ -1261,27 +1301,40 @@ public async Task GetResponseAsync_ResourceNotFound_ThrowsWebException()
...
             {
+                HttpWebRequest request = WebRequest.CreateHttp(uri);
+                request.ServerCertificateValidationCallback = (a, b, c, d) => true;
+                using WebResponse response = await request.GetResponseAsync();
                 string headersString = response.Headers.ToString();
-                string headersPartialContent = "Content-Type: application/json";
-                Assert.True(headersString.Contains(headersPartialContent));
-            }
+                Assert.Equal(HeadersPartialContent, response.Headers[HttpResponseHeader.ContentType]);
+            }, server => server.HandleRequestAsync(headers: new HttpHeaderData[] { new HttpHeaderData("Content-Type", HeadersPartialContent) }), options);
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Same echo server issue.  It seems like we should add a server.Echo() method that functionally does the same thing the remote echo server was doing.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 9 above)
Comment (ViktorHofer) [reply src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Opened https://github.com/dotnet/corefx/issues/38022

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 8 above)
Comment (ViktorHofer) [reply src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Calling EndGetResponse makes the test hang infinitely as it waits for a response which is never sent.

Comment (stephentoub) [reply src/System.Net.Requests/tests/HttpWebRequestTest.cs]: That's because you've written the test to not send back a response.  You would need to fix that, too.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 10 above)
Comment (davidsh) [reply src/System.Net.Requests/tests/HttpWebRequestTest.cs]: I agree with @stephentoub on this point. HttpWebRequest builds on HttpClient and we already have detailed tests for the HttpClient layer. So, we use these HttpWebRequest tests mainly to verify end-to-end functionality. And using an echo server helps satisfy that requirement.

**Comment by ViktorHofer** [comment]
> I suggest you revert the changes to the two problematic tests I commented on and get the rest of the changes merged

I implemented the echo functionality inline in the two tests as I thought it would be less work. After playing around with the server I noticed that the `connection.ReadToEndAsync` call doesn't read until the end but only the chunked part that was transmitted at that point. Instead I'm now reading from the stream directly...

**Comment by ViktorHofer** [comment]
All the tests are failing on UWP: `System.IO.IOException : Authentication failed because the remote party has closed the transport stream.`. Am I missing something? The other failures are unrelated.

**Comment by ViktorHofer** [comment]
Ping @davidsh, see question above

**Comment by davidsh** [comment]
>Ping @davidsh, see question above

I'm seeing errors in NETFX and UWP runs:

NETFX:
image

>Unhandled Exception of Type System.ArgumentException
Message :
System.ArgumentException : The 'Content-Type' header must be modified using the appropriate property or method.
Parameter name: name

And the UWP failures as well:
>All the tests are failing on UWP: System.IO.IOException : Authentication failed because the remote party has closed the transport stream.. Am I missing something? The other failures are unrelated.

I think the UWP failures are due to a limitation of trying to do loopback w/ TLS/SSL in the UWP environment.  The tests worked fine using a remote server with TLS. But loopback in UWP is always problematic due to running under app container. I'm not sure what the exact problem is though without investigating it.

I don't think there is value in spending a lot of time on this since it is some kind of test bug/limitation with UWP. So, I would recommend you add this to those tests to skip it on UWP:

```c#
[SkipOnTargetFramework(TargetFrameworkMonikers.Uap, "Loopback server with TLS has problems on UWP")]
```

**Comment by ViktorHofer** [comment]
Thanks for the help @davidsh and @stephentoub. We should update the docs that mention the LoopbackServer at some point.

**Comment by stephentoub** [comment]
> We should update the docs that mention the LoopbackServer at some point

Which docs?

**Comment by ViktorHofer** [comment]
https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/writing-tests.md#loopbackserver
